File: app/src/main/java/joebruckner/lastpick/managers/MovieManager.java
package joebruckner.lastpick.managers;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import info.movito.themoviedbapi.model.MovieDb;

public class MovieManager implements BulkMovieListener, SingleMovieListener {
	@Nullable private OnNewMovieListener listener;
	@NonNull private List<MovieDb> cachedResults;

	private static final String API_KEY = "9856c489f3e6e5b5b972c6373c440210";
	private static final String LANGUAGE = "en";
	private static final int PAGES = 3;
	private static final int SCOPE = 150;

	public MovieManager() {
		cachedResults = new ArrayList<>();
	}

	public void setOnNewMovieListener(@NonNull OnNewMovieListener listener) {
		this.listener = listener;
	}

	public void findNewMovie() {
		if (cachedResults.isEmpty())
			refillCache();
		else
			getMovie();
	}

	private void refillCache() {
		BulkMovieTask task = new BulkMovieTask(PAGES, SCOPE, API_KEY, LANGUAGE, this);
		task.execute();
	}

	private void getMovie() {
		Log.d(this.getClass().getSimpleName(), "Fetching Movie");
		int size = cachedResults.size();
		Random random = new Random(System.currentTimeMillis());
		int id = cachedResults.remove(random.nextInt(size)).getId();
		SingleMovieTask task = new SingleMovieTask(id, API_KEY, LANGUAGE, this);
		task.execute();
	}

	@Override public void onNewMovies(@NonNull List<MovieDb> movies) {
		this.cachedResults = movies;
		Log.d(this.getClass().getSimpleName(), "Cache Refilled");
		findNewMovie();
	}

	@Override public void onNewMovie(@NonNull MovieDb movie) {
		Log.d(this.getClass().getSimpleName(), "Fetched Movie");
		if (listener != null)
			listener.onNewMovie(movie);
	}

	@Override public void onError(@NonNull String errorMessage) {
		if (listener != null)
			listener.onError(errorMessage);
	}
}

File: app/src/main/java/joebruckner/lastpick/managers/MovieManager.kt
package joebruckner.lastpick.managers

import android.util.Log
import com.google.gson.JsonArray

import java.io.IOException

import info.movito.themoviedbapi.model.MovieDb
import joebruckner.lastpick.models.Movie
import joebruckner.lastpick.models.MovieSet
import org.json.JSONArray
import retrofit.Callback
import retrofit.GsonConverterFactory
import retrofit.Response
import retrofit.Retrofit

class MovieManager(var listener: OnNewMovieListener? = null) {
    val service = Retrofit.Builder()
            .baseUrl("")
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(TmdbService::class.java)

    fun findNewMovie() {
        //updateMovieSet()
        getMovieById(13)
    }

    private fun updateMovieSet(page: Int) {
        service.getTopRatedMovies(page).enqueue(object: Callback<MovieSet> {

            public override fun onResponse(response: Response<MovieSet>, retrofit: Retrofit) {
                if (response.isSuccess && response.code() == 200)
                    Log.d("IDs", getResultIds(response.body().results).toString())
                else
                    listener?.onError("Failed to fetch page $page of Top Rated")
            }

            public override fun onFailure(t: Throwable) {
                t.printStackTrace()
            }

        });
    }

    private fun getMovieById(id: Int) {
        service.getMovie(id).enqueue(object: Callback<Movie> {
            override fun onResponse(response: Response<Movie>, retrofit: Retrofit) {
                if (response.isSuccess && response.code() == 200)
                    Log.d("Movie", response.body().toString())
                else
                    listener?.onError("Failed to fetch Movie with Id $id")
            }

            override fun onFailure(t: Throwable) {
                t.printStackTrace()
            }

        })
    }

    private fun getResultIds(results: JsonArray):
            List<Int> = results.map { it.asJsonObject.get("id").asInt }

    companion object {
        private val LANGUAGE = "en"
        private val PAGES = 3
        private val SCOPE = 150
    }
}
